### PR TITLE
watch subscriptions in loadtest

### PR DESCRIPTION
**Pull request by k15r**
**Description**

Changes proposed in this pull request:

This pr changes the setup from the loadtester itself to the user. A user creates subscriptions and labels them with `eventing-loadtest: cloudevent` or `eventing-loadtest: legacy`. Those subscriptions will then be analyzed and events will be created according to their configuration.

- fix sending cloudevents
- introduce proper goroutine handling in the events
- synchronize all those go routines
- refactor sender

**Related issue(s)**

### CODE REVIEW

File: internal/loadtest/events/GenericEvent/events.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	"github.com/kyma-project/eventing-tools/internal/loadtest/events/payload"
+	"github.com/kyma-project/eventing-tools/internal/tree"
+)
+
+type Event struct {
+	source    string
+	version   string
+	EventName string
+	eps       int
+	Starttime string
```
Comment (grischperl) [inline internal/loadtest/events/GenericEvent/events.go:21]: What is the reason for `EventName` and `Starttime` being exported?

File: internal/loadtest/publisher/publisher.go [hunk 2]
```diff
@@ -1,40 +1,57 @@
...
 
-func Start() {
+func Start(port int) {
 	appConfig := config.New()
 	k8sConfig := k8s.ConfigOrDie()
 	k8sClient := k8s.ClientOrDie(k8sConfig)
+	dynamicClient := dynamic.NewForConfigOrDie(k8sConfig)
 
 	legacyEventSender := legacyevent.NewSender(appConfig)
+	legacySender := sender2.NewSender(appConfig, legacyEventSender)
```
Comment (grischperl) [inline internal/loadtest/publisher/publisher.go:27]: Naming: `legacyevent.NewSender` returns a type `Sender`
`sender2.NewSender` returns a type `EventSender`

Wouldn't naming make more sense if:
```
legacySender := legacyevent.NewSender
legacyEventSender := sender2.NewSender
```

File: internal/loadtest/publisher/publisher.go [hunk 3]
```diff
@@ -1,40 +1,57 @@
...
-		OnDeleteNotify(infraInstance).
-		OnAddNotify(legacyEventSender).
-		OnUpdateNotify(legacyEventSender).
-		OnDeleteNotify(legacyEventSender).
-		OnAddNotify(cloudEventSender).
-		OnUpdateNotify(cloudEventSender).
-		OnDeleteNotify(cloudEventSender).
+		// OnAddNotify(infraInstance).
+		// OnUpdateNotify(infraInstance).
+		// OnDeleteNotify(infraInstance).
```
Comment (grischperl) [inline internal/loadtest/publisher/publisher.go:35]: Are these expected to be needed again?

File: internal/loadtest/publisher/publisher.go (hunk 2 above)
Comment (k15r) [reply internal/loadtest/publisher/publisher.go:27]: absolutely :D

File: internal/loadtest/events/GenericEvent/events.go (hunk 1 above)
Comment (k15r) [reply internal/loadtest/events/GenericEvent/events.go:21]: changed

File: internal/loadtest/publisher/publisher.go (hunk 3 above)
Comment (k15r) [reply internal/loadtest/publisher/publisher.go]: ... and gone

File: docs/loadtest/README.md [hunk 4]
```diff
@@ -5,36 +5,23 @@ This tool is used to generate continues load on the Eventing components.
...
- - EPS **50**: `order.created.v0050`.
- - EPS **90**: `order.created.v0090
+> Note: Encoding the EPS in the event type version is used only for debugging purposes and is not a production use-case.
 
-*Cloud Events* have the `eps` encoded at the end as part of the `version`.
-*Legacy Events* have the `eps`  body in the `event-type` field of the body.
+> There is no special handling of subscriptions otherwise. The `sink` and also the `maxInFlight` settings will be respected.
 
-> Note: Encoding the EPS in the event type version is used only for debugging purposes and is not a production use-case.
+This publisher works best in combination with the loadtest-subscriber. Two instances of the subscriber are deployed using the kustomize deployment. Configure the subscriptions to use those by setting the sink to `loadtest-subscriber-0.eventing-test.svc.cluster.local` or `loadtest-subscriber-1.eventing-test.svc.cluster.local`
```
Comment (marcobebway) [inline docs/loadtest/README.md:24]: Do we still need to have 2 sinks? Does it make sense to always use the same sink and delete the other?

Comment (k15r) [reply docs/loadtest/README.md:24]: the two sinks are configured to:
- 1. run on the same node as ec
- 2. run on different node than ec

but i agree... the whole concept of those sinks should be reconsidered

File: docs/loadtest/README.md [hunk 5]
```diff
@@ -5,36 +5,23 @@ This tool is used to generate continues load on the Eventing components.
 It does that by sending Cloudevents to the Eventing publisher proxy forever and consume them inside Kyma functions.
 Ideally, it should be used when JetStream is used as the active Eventing backend.
 
-## Eventing infrastructure
+## Subscriptions
 
-For testing against **Legacy Events** set `use_legacy events` to `true`. Leave it to `false` to test against **Cloud Events**.
+The publisher watches all subscriptions in the cluster.
+To activate sending events for one subscription you have to label that subscription with one of these labels:
```
Comment (NHingerl) [inline docs/loadtest/README.md:11]:
```suggestion
To activate sending events for one subscription, you must label that subscription with one of the following labels:
```

File: docs/loadtest/README.md [hunk 6]
```diff
@@ -5,36 +5,23 @@ This tool is used to generate continues load on the Eventing components.
...
+- eventing-loadtest: legacy 
+- eventing-loadtest: cloudevent
 
-The load test will create:
+The load tester will then create events using the event types configured in the subscription and send them with the format configured in the label.
```
Comment (NHingerl) [inline docs/loadtest/README.md:15]:
```suggestion
After you configured the event type in the label, the load tester creates events accordingly and sends them with the format configured in the label.
```

File: docs/loadtest/README.md [hunk 7]
```diff
@@ -5,36 +5,23 @@ This tool is used to generate continues load on the Eventing components.
...
 
-1. **Two** Kyma function subscribers.
-2. **Two** Kyma subscriptions.
+The load generated per event type will be extracted from the event type. To do this the event type must match the following pattern:
```
Comment (NHingerl) [inline docs/loadtest/README.md:17]:
```suggestion
The load generated per event type is extracted from the event type. To do this, the event type must match the following pattern:
```

File: docs/loadtest/README.md [hunk 8]
```diff
@@ -5,36 +5,23 @@ This tool is used to generate continues load on the Eventing components.
...
+`<event-name>.v<Number of events per second>`, eg. `order.created.v500` will create events of type order.create.v500 and the sender will try to publish 500 events per second.
```
Comment (NHingerl) [inline docs/loadtest/README.md:18]:
```suggestion
`<event-name>.v<Number of events per second>`. For example, `order.created.v500` creates events of type order.create.v500 and the sender tries to publish 500 events per second.
```

**Review by NHingerl — COMMENTED** [review]
came across this PR and did minor language editing :)

**Review by marcobebway — COMMENTED** [review]
To improve the user experience, does it make sense to have make targets that builds the following?
- eventing-tools/resources/loadtest/ko
- eventing-tools/resources/loadtest/pr

**Review by marcobebway — APPROVED** [review]

**Review by NHingerl — APPROVED** [review]

**Review by marcobebway — APPROVED** [review]

**Pull request merged**